cpp/src/arrow/memory_pool.cc

```diff
@@ -603,14 +643,109 @@ class BaseMemoryPoolImpl : public MemoryPool {
     stats_.UpdateAllocatedBytes(-size);
   }
 
-  void ReleaseUnused() override { Allocator::ReleaseUnused(); }
+ protected:
+  virtual Status AllocateImmutableZeros(int64_t size, uint8_t** out) {
+#ifdef USE_MMAP_FOR_IMMUTABLE_ZEROS
+    if (size > 0) {
+      *out = static_cast<uint8_t*>(mmap(
+          nullptr, size, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE, -1, 0));
```

Comment:
I'm not sure this is useful, because the underlying allocator probably already relies on mmap for large-ish allocations.

Reply:
Ah, you're right, we don't know that the allocator returned us zero-allocated memory, unfortunately.
(note that both jemalloc and mimalloc have zero-initializing allocation APIs, so we could use those)

Reply:
In fact, `Allocate()` generally does not return zero-initialized memory, due to the poisoning mechanic. I would rather not dive into this rabbit hole as well right now, so I just added a TODO comment in the code. I can file a followup JIRA for it too, if you want.

Reply:
@jvanstraten You're right. So let's keep it like this and just explain why it is in a comment.

Reply:
Note that [VirtualAlloc](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc) on Windows zero-initializes pages. We can fall back on `calloc` by default.

Reply:
I think 3738819 does what you want in terms of refactoring? I left the CMake option to disable the OS-specific stuff and use the regular memory pool allocator in, so people can still opt to use that allocator (and the statistics tracking thereof) in favor of `calloc()`.

I don't have a Windows dev environment set up though, so I'd prefer to open a followup issue for VirtualAlloc instead, so someone else can pick it up if they want. Same for mac; I guess mmap will probably work, but I can't test it.